client/internal/retry/do.go

```diff
@@ -142,3 +144,84 @@ func Do(retryFunc Func, opt Options) error {
 		errors.Join(accumulatedErrs...),
 	)
 }
+
+func Do2(ctx context.Context, retryFunc Func, opt Options) error {
+	// set up context
+	ctxToUse := ctx
+	if opt.Timeout > 0 {
+		var cancel context.CancelFunc
+		ctxToUse, cancel = context.WithTimeout(ctx, opt.Timeout)
+		defer cancel()
+	}
+
+	// set up channels
+	retryChan := make(chan retryInfo)
+	go retry(ctxToUse, retryChan, opt.Delay, opt.Retries)
+	return doInternal(ctxToUse, retryFunc, retryChan, opt)
+}
+
+type retryInfo struct {
+	attempt int
+	timeout bool
+}
+
+// retry sends a retryInfo to the given channel according to given delay and retries,
+func retry(ctx context.Context, c chan<- retryInfo, delay time.Duration, retries int) {
+	c <- retryInfo{attempt: 0, timeout: false}
+	for attempt := 1; attempt <= retries || retries < 0; attempt++ {
+		if willTimeoutAfterDelay(ctx, delay) {
+			c <- retryInfo{attempt: attempt, timeout: true}
+		}
+		time.Sleep(delay)
+		c <- retryInfo{attempt: attempt, timeout: false}
+	}
+	close(c)
+}
+
+func willTimeoutAfterDelay(ctx context.Context, delay time.Duration) bool {
+	ddl, ok := ctx.Deadline()
+	if !ok {
+		return false // no deadline is set, so we will never time out after delay
+	}
+	return time.Now().Add(goutil.Max(delay, 0)).After(ddl)
+}
+
+func doInternal(ctx context.Context, retryFunc Func, retryChan <-chan retryInfo, opt Options) error {
+	// setup time
+	startTime := time.Now()
+	// setup errors
+	// retryErrors[i] = nil: no error occur at this attempt
+	// retryErrors[i] != nil: error occur at this attempt
+	retryErrors := make([]error, goutil.Max(opt.Retries, 0)+1)
+
+	// setup retry info
+	var retryInfo retryInfo
+	var retryOk = true
+
+	for {
+		select {
+		case <-ctx.Done():
+			// context timeout
+			return fmt.Errorf("%w at attempt=%d/%d, after=%s, retry errors=%w", ctx.Err(), retryInfo.attempt, opt.Retries, time.Since(startTime), errors.Join(retryErrors...))
+		case retryInfo, retryOk = <-retryChan:
```

Comment:
I think using channels here makes the code difficult to follow. 
We can use a standard for-loop with a nested select 

So for example (warning! psuedo code):

```go
for attempt := 1; attempt <= retries; attempt++ {
    select {
        case <- ctx.Done():
            return fmt.Error("uh oh")

        default:
            retryFunc()
        }
}
```

This removes the mental overhead of messages being passed between channels.

Does this make sense?
Have I interpretted your code correctly?

Reply:
Discussed offline, thanks Chris 👍